src/ice/mod.rs

```diff
@@ -1146,4 +1226,171 @@ mod test {
+                }
+                NatType::PortRestrictedCone { mappings } => {
+                    // For a port-restricted-cone NAT, the key is also the 4-tuple of the connection.
+                    // Contrary to a symmetric NAT, the source port is preserved.
+                    *mappings.entry((from, to)).or_insert(from.port())
+                }
+            };
+
+            (SocketAddr::new(self.external_ip, external_port), to)
+        }
+
+        fn transform_inbound(
+            &self,
+            from: SocketAddr,
+            to: SocketAddr,
+        ) -> Option<(SocketAddr, SocketAddr)> {
+            let internal_addr = match &self.nat_type {
+                // For symmetric NAT, need exact source match of from == dst and only on the assigned port.
+                NatType::Symmetric { mappings } => {
+                    let ((src, _), _) = mappings
+                        .iter()
+                        .find(|((_, dst), outside)| from == *dst && **outside == to.port())?;
+
+                    *src
+                }
+                // For restricted-cone NAT, traffic on the outside assigned port is routed back
+                // if we have previously contacted this IP + port combination.
+                NatType::PortRestrictedCone { mappings } => {
+                    let ((src, _), _) = mappings
+                        .iter()
+                        .find(|((_, dst), outside)| from == *dst && **outside == to.port())?;
+
+                    *src
+                }
+            };
+
+            Some((from, internal_addr))
+        }
+    }
+
+    fn symmetric_nat_lookup(
+        src: SocketAddr,
+        dst: SocketAddr,
+        mappings: &mut HashMap<(SocketAddr, SocketAddr), u16>,
+    ) -> u16 {
+        let outside_port = loop {
+            let port = rand::random();
+
+            if mappings.values().any(|p| *p == port) {
+                continue;
+            }
+
+            break port;
+        };
+
+        *mappings.entry((src, dst)).or_insert(outside_port)
+    }
+
+    /// Transform function with flexible NAT support
+    fn transform(
+        from: SocketAddr,
+        to: SocketAddr,
+        from_agent: &mut TestAgent,
+        to_agent: &mut TestAgent,
+    ) -> Option<(SocketAddr, SocketAddr)> {
+        if from.port() == 9999 || to.port() == 9999 {
+            return None;
+        }
+
+        let outgoing_is_from_relay = from_agent
+            .local_candidates()
+            .iter()
+            .any(|c| c.addr() == from && c.kind() == CandidateKind::Relayed);
+
+        // If NAT is present on sending agent, apply it.
+        let (from, to) = if let Some(nat) = &mut from_agent.nat {
+            // If our traffic is "from" a relay candidate, the NAT does not apply.
+            if outgoing_is_from_relay {
+                (from, to)
+            } else {
+                nat.transform_outbound(from, to)
+            }
+        } else {
+            (from, to)
+        };
```

Comment:
let-chains would make this a lot nicer but we also run our tests for our MSRV so we can't use it here :(

Reply:
Yeah, it's ok. One day we raise MSRV :)